Plugins/Events/Events/ResourceEvents.cpp

```diff
@@ -0,0 +1,161 @@
+#include "Events/ResourceEvents.hpp"
+#include "API/CExoBase.hpp"
+#include "API/CExoAliasList.hpp"
+#include "API/CNWSModule.hpp"
+#include "API/CExoResMan.hpp"
+#include "API/Functions.hpp"
+#include "API/Constants.hpp"
+#include "Events.hpp"
+#include "Utils.hpp"
+
+#include <sys/inotify.h>
+#include <unistd.h>
+#include <poll.h>
+#include <climits>
+#include <fcntl.h>
+
+namespace Core {
+    extern bool g_CoreShuttingDown;
+}
+
+namespace Events {
+
+using namespace NWNXLib;
+using namespace NWNXLib::API;
+using namespace NWNXLib::API::Constants;
+
+#define EVENT_SIZE          ( sizeof (struct inotify_event) )
+#define EVENT_BUF_LEN       ( 128 * ( EVENT_SIZE + NAME_MAX + 1) )
+
+static bool processEvents;
+static int selfPipeFds[2];
+std::unique_ptr<std::thread> ResourceEvents::m_pollThread;
+
+ResourceEvents::ResourceEvents(Services::TasksProxy* tasks)
+{
+    Events::InitOnFirstSubscribe("NWNX_ON_RESOURCE_.*", [tasks]()
+    {
+        int ret = pipe2(selfPipeFds, O_NONBLOCK);
+
+        if (ret == -1)
+        {
+            LOG_ERROR("Failed to initialize self-pipe with error: %s", std::strerror(errno));
+            return;
+        }
+
+        int pipeReadFd = selfPipeFds[0];
+        CExoString exoPath = Globals::ExoBase()->m_pcExoAliasList->GetAliasPath("NWNX");
+        std::string path = exoPath.CStr();
+
+        m_pollThread = std::make_unique<std::thread>([tasks, path, pipeReadFd]()
+        {
+            struct pollfd fds[2];
+            char buffer[EVENT_BUF_LEN];
+
+            fds[0].fd = pipeReadFd;
+            fds[0].events = POLLIN;
+
+            int fd = inotify_init1(IN_NONBLOCK), wd = 0;
+
+            if (fd != -1)
+            {
+                fds[1].fd = fd;
+                fds[1].events = POLLIN;
+
+                wd = inotify_add_watch(fd, path.c_str(), IN_CREATE | IN_DELETE | IN_MODIFY);
+
+                if (wd != -1)
+                {
+                    processEvents = true;
+                }
+            }
+
+            auto ProcessEvent = [tasks](const std::string &eventType, const std::string &file) -> void
+            {
+                tasks->QueueOnMainThread([eventType, file]()
```

Comment:
Fun fact: NWN can take up to a second to actually register the resource after it's been copied in (due to lack of cross-platform inotify API). So it's very likely that on any event here you will actually see the stale data if you attempt to read the resource.

The easy way to solve this would be to just delay queuing here by 1 second. A better way might be to push all these events into a buffer, and also hook `CExoKeyTable::CheckForModificationsInStorage`. That thing runs every second when NWN checks the folders. After it has run, you can queue the event on the main thread and it'll guarantee the resources are visible.

Reply:
I did the easy way and just sleep the thread for a second, how would the better way best be done?

Reply:
> how would the better way best be done?

1) Define an event structure that contains `enum { ADDED, REMOVED, MODIFIED} type; string resource;`. Define a global `vector<eventstruct>` that stores all events.
2) shared-hook `CheckForModificationInStorage`. If your global vector is not empty, `QueueOnMainThread` a function that signals the event, once for every vector element, and then clears the vector.

This is still a bit racey, so _ideally_ there is a binary semaphore that your shared hook acquires when it runs the QueueOnMainThread, and the main thread lambda releases it when all events are signaled. Since that's a hassle, it's enough to just use a `std::mutex` and acquire it in the main thread lambda when iterating over the array, as well as the side thread when inserting into it.

All that said, this works well enough for me, so if you just wanna merge, go ahead.

Reply:
I'll give the shared hook a try sometime and do a followup PR if that works well